modules/userId.js

```diff
@@ -0,0 +1,431 @@
+      localStorage.setItem(storage.name, encodeURIComponent(valueStr));
+    }
+  } catch (error) {
+    utils.logError(error);
+  }
+}
+
+/**
+ * @param {SubmoduleStorage} storage
+ * @returns {string}
+ */
+export function getStoredValue(storage) {
+  let storedValue;
+  try {
+    if (storage.type === COOKIE) {
+      storedValue = utils.getCookie(storage.name);
+    } else if (storage.type === LOCAL_STORAGE) {
+      const storedValueExp = localStorage.getItem(`${storage.name}_exp`);
+      // empty string means no expiration set
+      if (storedValueExp === '') {
+        storedValue = localStorage.getItem(storage.name);
+      } else if (storedValueExp) {
+        if ((new Date(storedValueExp)).getTime() - Date.now() > 0) {
+          storedValue = decodeURIComponent(localStorage.getItem(storage.name));
+        }
+      }
+    }
+    // we support storing either a string or a stringified object,
+    // so we test if the string contains an stringified object, and if so convert to an object
+    if (typeof storedValue === 'string' && storedValue.charAt(0) === '{') {
+      storedValue = JSON.parse(storedValue);
+    }
+  } catch (e) {
+    utils.logError(e);
+  }
+  return storedValue;
+}
+
+/**
+ * test if consent module is present, applies, and is valid for local storage (purpose 1)
+ * @param {Object} consentData
+ * @returns {boolean}
+ */
+export function hasGDPRConsent(consentData) {
+  if (consentData && typeof consentData.gdprApplies === 'boolean' && consentData.gdprApplies) {
+    if (!consentData.consentString) {
+      return false;
+    }
+    if (consentData.vendorData && consentData.vendorData.purposeConsents && consentData.vendorData.purposeConsents['1'] === false) {
+      return false;
+    }
+  }
+  return true;
+}
+
+/**
+ * @param {Object[]} submodules
+ * @param {function} [processCompleted] - not required, executed when all callbacks have returned responses
+ */
+export function processSubmoduleCallbacks(submodules, processCompleted) {
+  submodules.forEach(function(submodule) {
+    submodule.callback(function callbackCompleted (idObj) {
+      // clear callback, this prop is used to test if all submodule callbacks are complete below
+      submodule.callback = undefined;
+
+      // if valid, id data should be saved to cookie/html storage
+      if (idObj) {
+        setStoredValue(submodule.config.storage, idObj, submodule.config.storage.expires);
+
+        // cache decoded value (this is copied to every adUnit bid)
+        submodule.idObj = submodule.submodule.decode(idObj);
+      } else {
+        utils.logError(`${MODULE_NAME}: ${submodule.submodule.name} - request id responded with an empty value`);
+      }
+
+      // Done when every submodule callback is set to 'undefined'
+      if (submodules.every(item => typeof item.callback === 'undefined')) {
+        // Notify done through calling processCompleted
+        if (typeof processCompleted === 'function') {
```

Comment:
more efficient to check this first before checking the callback in every submodule.